File: RFID/sensores.py
import RPi.GPIO as GPIO
import time
import signal
import sys

# Configurações de hardware
GPIO.setmode(GPIO.BOARD)
TRIG_PINS = [18, 16, 12]
ECHO_PINS = [24, 26, 22]
LED_VERDE_PINS = [20, 19, 13]
LED_VERMELHO_PINS = [21, 17, 27]

for pin in TRIG_PINS:
    GPIO.setup(pin, GPIO.OUT)
for pin in ECHO_PINS:
    GPIO.setup(pin, GPIO.IN)
for pin in LED_VERDE_PINS + LED_VERMELHO_PINS:
    GPIO.setup(pin, GPIO.OUT)

def medir_distancia(trig_pin, echo_pin):
    # Envia o pulso de trigger
    GPIO.output(trig_pin, True)
    time.sleep(0.00001)
    GPIO.output(trig_pin, False)

    # Calcula o tempo de ida e volta do sinal ultrassônico
    while GPIO.input(echo_pin) == 0:
        inicio = time.time()
    while GPIO.input(echo_pin) == 1:
        fim = time.time()

    duracao = fim - inicio
    distancia = (duracao * 34300) / 2  # velocidade do som (34300 cm/s)

    return distancia

# Função para finalizar o programa
def finalizar_programa(signal, frame):
    print("\nFinalizando o programa...")
    GPIO.cleanup()
    sys.exit(0)

# Captura o sinal de interrupção (CTRL+C)
signal.signal(signal.SIGINT, finalizar_programa)

# Loop principal
try:
    while True:
       for i in range(3):
            distancia = medir_distancia(TRIG_PINS[i], ECHO_PINS[i])
            print(f"Sensor {i + 1} - Distância: {distancia:.2f} cm")

            # Aciona LEDs com base na presença
            if distancia < 30:  # Ajuste para a distância desejada em cm
                GPIO.output(LED_VERDE_PINS[i], True)
                GPIO.output(LED_VERMELHO_PINS[i], False)
                print(f"Sensor {i + 1} - Presença detectada - LED Verde ON")
            else:
                GPIO.output(LED_VERDE_PINS[i], False)
                GPIO.output(LED_VERMELHO_PINS[i], True)
                print(f"Sensor {i + 1} - Ausência detectada - LED Vermelho ON")

            time.sleep(5)  # Define o intervalo entre as leituras

finally:
    GPIO.cleanup()

File: RFID/sensores.cpp
const int TOTAL_SENSORES = 3; // Total de sensores
const int TRIG_PINS[] = {3, 4, 5};
const int ECHO_PINS[] = {6, 7, 8};
const int LED_VERDE_PINS[] = {9, 10, 11};
const int LED_VERMELHO_PINS[] = {12, 13, A0};
const int DISTANCIA_THRESHOLD = 10; // Limite para detecção de presença

void setup() {
  Serial.begin(9600);
  for (int i = 0; i < TOTAL_SENSORES; i++) {
    pinMode(TRIG_PINS[i], OUTPUT);
    pinMode(ECHO_PINS[i], INPUT);
    pinMode(LED_VERDE_PINS[i], OUTPUT);
    pinMode(LED_VERMELHO_PINS[i], OUTPUT);
  }
}

long medirDistancia(int trigPin, int echoPin) {
  digitalWrite(trigPin, LOW);
  delayMicroseconds(2);
  digitalWrite(trigPin, HIGH);
  delayMicroseconds(10);
  digitalWrite(trigPin, LOW);
  long duracao = pulseIn(echoPin, HIGH);
  return duracao * 0.034 / 2; // Distância em cm
}

void controlarLEDs(int indiceSensor, bool presencaDetectada) {
  if (presencaDetectada) {
    digitalWrite(LED_VERDE_PINS[indiceSensor], LOW);
    digitalWrite(LED_VERMELHO_PINS[indiceSensor], HIGH);
  } else {
    digitalWrite(LED_VERDE_PINS[indiceSensor], HIGH);
    digitalWrite(LED_VERMELHO_PINS[indiceSensor], LOW);
  }
}

void enviarDadosParaPython(int sensoresAtivos) {
  int sensoresInativos = TOTAL_SENSORES - sensoresAtivos;
  Serial.println(sensoresInativos);
}


void loop() {
 
  int sensoresAtivos = 0; // Reseta o contador de sensores ativos para calcular novamente

  for (int i = 0; i < TOTAL_SENSORES; i++) {
    long distancia = medirDistancia(TRIG_PINS[i], ECHO_PINS[i]);
    Serial.print("Sensor ");
    Serial.print(i + 1);
    Serial.print(" - Distância: ");
    Serial.print(distancia);
    Serial.println(" cm");

    bool presencaDetectada = (distancia < DISTANCIA_THRESHOLD);
    if (presencaDetectada) {
      sensoresAtivos++;
    }

    controlarLEDs(i, presencaDetectada);
  }

  enviarDadosParaPython(sensoresAtivos);

  delay(5000); // Atraso antes da próxima leitura
}

